### ISSUE TITLE
P0009r8 Post-SanDiego mailing

### ISSUE THREAD

**Issue by mhoemmen** [ORNL/cpp-proposals-pub#84]
Per request by @hcedwar, I've attached the minutes from Saturday's LWG small group review of P0009R8 at San Diego.
```
---+ Saturday LWG: P0009R8 (mdspan)

Morning, after Plenary.

Present

   * Damien Lebrun-Grande
   * Jorg Brown
   * Carter Edwards
   * Alasdair Meredith
   * Dan Sunderland
   * Mark Hoemmen (scribe)
   * Geoff Romer (chair)

CE: Paper as it stood in the pre-meeting mailing.

DS: Targeting Library Fundamentals TS3.  Wording changed substantially since Rapperswil, based on feedback from there.  Prefer to go over all wording, since diffs large.

GR: What's in views?

MH: span.

DS: string_ref?

GR: Not unless they've moved it.

GR: "Add the following paragraphs to [views.general]" needs to say _where_ in that section it's added.  It currently consists of one paragraph.  Would be good to depict this as the existing section with green text for insertion, just for context.  Helpful to have that when editing something that exists.

[bikeshed can do markup; mdspan didn't diff from last time since too many changes]

GR: This seems like the wrong place for that text: "The basic_mdspan object ... is a span of elements."  Detailed explanation of mdspan that would be separated by lots of text.  Move to before synopsis of mdspan.  Same for this (the subspan function).  Only first sentence of that paragraph should stay in [views.general].

DS: OK.

GR: "A reference of an element" -- should that be "reference to an element"?

DS: Yes.

GR: Text defines words of power with italics, and also boldface.

DS: Was meant to be italics; markup error; we'll fix that.

DR: Right now, we have multidimensional index, redefine as multindex.  Should both be italics or only one?

GR: You should pick one term and use it.

DS: Let's strike multi-index and use multidimensional index consistently.

GR: OK.  We don't usually optimize the standard for brevity.

[AM arrives in the room.]

[resolved: reorder to define terms before first use]

GR: Nice to have this (2nd) overview sentence before just diving into defs, but would be preferable to define things in order.

DS: Overview sentence did not italicize words we would define later on.

GR: Yes; words of power only italicized where they are def'd.

GR: Replace sentence 2 with something more colloquial, doesn't rely on terms like codomain.  We know what multidimensional index means.  Drop "within its domain."

MH,CE: "... to a reference to an element."?

GR: This should mention more prominently the fact that basic_mdspan is a view.  The fact that it wraps some other container that it does not own.

CE: Say "nonowning" in there?

GR: Yes, probably.

DS: We can wordsmith that on reflector.

-------

GR: Paragraph 2: Don't italicize domain.

DS: In general, review italicization of words of power.

[OK]

GR: Could strike para 2 entirely, or move to front matter of subspan function.

DS: The latter.

---++ Header synopsis

GR: This is targeting the TS?

DS: Yes.

AM: Inline namespace std::experimental::fundamentals_v3, instead of having 3 levels.

GR: Generally, stuff in TS just goes in experimental.  I don't think we've usually had a fundamentals_v3 namespace.  Doublecheck whether this should be an inline namespace.  Namespace is fine, but make fundamentals_v3 an inline namespace.  Because this is based on 17, can't use "::" for inline namespace.

JB: General guidelines for use of std::experimental namespace?

AM: All TSes for library features intending to go into std namespace, will have inlined namespace beyond that which will allow for versioning.  If landing into a namespace like ranges, namespace experimental, ranges, inline version number.

DS: On accessor policy, don't need extra whitespace on "> >".

GR: Can we do ... on non-template type parameter?

CE,AM: Yes, since 11.

GR: OK.

AM: Hard to implement tuple if you can't, which is why we've needed it since 11.

DS: We should fix the line wrapping in synopsis.

AM: Conform: wide vs. narrow contract, sizes of the extents the same. No run-time checks here; purely compile time?

DS: There could be a run-time check if one of those extents is marked as dynamic.  It has a well-defined result in that case.

AM: OK.

AM: How much has changed since the last time we reviewed it.

DS: Of the synopsis, not much, of the actual wording, all of it.  We accounted for five hours of feedback.

GR: Exposition only mdspan_subspan -- the whole struct?

DS,CE: Yes.

GR: So this type is not guaranteed to exist?

DS: Correct. Where should comment go?

AM: Should we take exposition-only declarations out of synopsis altogether?

GR: I agree.  Such types may appear in text below.

DS: We do have exposition-only types in ranges, right?

AM: They have exposition-only concepts [not types].

MH,CE,DS: We use exposition-only type mdspan_subspan to define the return type of subspan.

GR: The usual way we do that, is "see below" in italics, for the return type.

DR: That would give us a place to do the exposition-only thing.

[OK, it returns "see below" as the return type, in italics.]

GR: Think about, even with this exposition-only type moved down, I think these names (e.g., "mdspan_subspan") are likely to give too much of an impression that they are actual library types.  There's one place in library where we have an actual exposition-only class.

DS: In exposition-only concepts, we used "kabob-case"; woudl that work?

GR: istream_buf_iterator_proxy

AM: Fact that it's a proxy calls out ...

GR: And also that it's a class member ....

AM: Why exposition only?

MH: Freedom for QOI

AM: It's a method function to compute that name for the result type.

DS,CE: Yes.

GR: Return type should be "basic_mdspan<see below>" with "see below" in italics, non code font.

GR: Best practices for tag types have moved around a bit.  There's an issue with ... default constructor is explicit, at least.  nullop_t is specified not to have a defult ctor or initializer_list ctor, and is spec'd not to be an aggregate.  Take a note to look off line into why nullop_t is specified the way it is.

DS: I want to raise issue of name "all".

CE: LEWG was fine with name "all".

GR: Which astonishes me.

AM: Nice vocabulary as long as it's only this thing.

DS: It might belong elsewhere / be useful for other things.

GR: Did LEWG discuss this?

DS: I was not in the room.

CE: Would have to go back to notes.

GR: Make sure that LEWG has considered whether they want a tag with this name in this header.

---++ mdspan.extents

MH: Same issue w/ bold and italics.

DS: And repeat definition.

GR: I would rather have it [definition] here than in front matter.

DS: Clean up front matter to be more colloquial.

MH: "i-th"?

GR: Not always consistent, I think usually not code font.

MH: Italics make it hard to see.

GR: Yes.

GR: Para 4: "is expected to store only" dynamic extents.

DS: Yes.

CE: Then you don't need a note.

GR: Perhaps whole thing should be a note.

GR: "should not" is magic wording for nonnormative encouragement.

AM: Do we need to say anything at all?

MH: Does not matter for correctness?

CE: Does matter for QOI.

GR: Does not matter for correctness, e.g., passing by ref or value.  Thus, I would encourage you to strike it.

CE: Move it all to a Note.

AM: I think, just strike it.  Having it there invites quibbling on wording.

AM: It sounds like DynamicRank and DynamicIndex are being used to define each other circularly, in para 3.

MH: They are just counting the number of -1's, right?

[discussion]

AM: DynamicRank denotes an index.  Why is an index a rank?

JB: DynamicIndex doesn't appear anywhere else in this paper.

DS: Could be artifact.

GR: Could be off by one bug in definition of DynamicIndex, but if not used, problem solved.

GR: I think, if DynamicIndex goes away, I don't care so much whether this is called rank or index.

MH: We did talk about span analogy of para 5, Constraints vs. Mandates.

GR: This is fine; we don't have anything like that for classes.

DS: Same thing with namespaces.

DS: There is no "size" method on mdspan; span issue does not arise.

[OK]

AM: Nicely sidestepped [potential for confusion].

AM: If I call static_extent with out-of-bounds index, what happens?

CE: It returns 1.

AM: Well-defined values, not a precondition; that's all I care about.  Same for extent.

[OK; no change needed]

CE: I see; Exposition-only stuff led DynamicIndex go away.

AM: Can't we just say value-initializes dynamic_extents_?  What you're trying to do is value-initialize it.  Unnecessary to say aggregate....

GR: "value-initializes dynamic_extents_" -- clearer to a reader that that's initializing it to zeros, not indeterminate values.

GR: I don't think we use that for-all character in the standard; I think we use words.  "For all r in the range [0,R)..."

AM: I'll float a crazy that this is a TS and open to experimentation.

MH: We didn't intend it as an experiment.

DS: I like the notation.

MH: I don't; less accessible to those using screen readers.

GR: Authors' choice; be consistent.  If you change it to words, you won't get asked to revise it.

AM: If you keep it, make a note that you intend this as an experiment.

CE: This is the place to experiment with it, because this is an intensely mathematical class, so let's try some mathematical notation.

GR: Where's dynamic_extent defined?  [while looking at extents() default ctor]

CE: In span.

GR: OK, got it.

GR: extent(r) is a member function?

DS: Yes.

GR: I would say, "this->extent(r)", for clarity.  Not entirely clear that the wording here is notionally evaluated within the function body.

CE: Should we do that for the rank() and static_extent() member functions as well?

GR: I assume rank() is just sizeof(extents)?

CE: Yes.

GR: Yes, that's my suggestion.

dynamic_extent is defined int he header of std::span.

JB: There is some movement to get std::span out of standard in favor of ranges.

[span is not 17]

AM: I'm proposing a paper rebasing span onto 20.  That would let you use span and its material without explicit wording pulling it in.

JB: Do we need a move constructor?

CE: For completeness, but not necessary for functionality.

GR: It can always be added as QOI.

JB: Looks like extent only ever had an array of integers.

DS: Correct.

GR: Then strike the move constructor.  Don't delete them, but don't declare them.

GR: [extents copy constructor] "Copy-initializes dynamic_extents_ from other.dynamic_extents_".

GR: I'm wavering on my feedback that there should be "this" [in "this->extent(r)"].

DS: If a member function, we weren't calling it from this, but if we were trying to pick it up from a different namespace...

GR: That's a separate issue.

AM: If a member function comes from a dependent base, need this, else don't.

GR: I agree that there's no need for it in code.  Maybe OK to omit "this".  I'll withdraw that feedback [above] unless someone else agrees with me.

GR: Does extents have an equality comparison?

DS: Yes.

GR: Could this just be "*this == other"?  [in copy ctor of extents]

DS: Yes.

GR: Do that.

[operator== is a nonmember function]

GR: Are the two Ensures conditions in default ctor two diff ways of saying the same thing?

CE: No.

CE: The Effects and the Ensures actually state the same thing.  Ensures was there before the exposition-only member was added to clarify.  It is redundant now.

AM: Given that we're not defaulting the copy constructor, we probably want to put noexcept on the copy constructor.

GR: std::array is allowed to be size zero, right?

AM: Yes.

AM: UB to ever call [] or front or back on a zero-length std::array, but they are declared.

GR: "Strike the Ensures, not the Effects" for default ctor.

[break for lunch]

-------

Afternoon.

Present:

   * Jorg Brown
   * CE
   * C. W. Holeman II
   * Vincent Reverdy
   * AM
   * DS
   * MH
   * GR

GR: In copy ctor of extents, drop redundant Ensures, as discussed above.

CE: OK

CE: Bring down noexcept from header.

DS: AM mentioned we need noexcept on the copy, since it acts as a move constructor as well (once we remove move constructor).

GR: Requires clause, logically prior to construction.

DS: Might be better as Ensures.

AM: Requires -> Constraints, Mandates, or Expects.

GR,AM: This is run time, so Expects.

GR: Phrasing in terms of member call to object ...

CE: It's a static member function; does that matter?

[C. W. Holman arrives]

GR: Yeah, OK ... wording this in terms of the template parameters would be annoying.

CE: Yes, that's what DynamicIndex was for, etc.

AM: For dynamic extent(s), we're making no requirements, but static extents must match.

GR: Yes.

AM: That's statically determinable, so mandates.

DS: It's run time.

GR: This lets you convert dynamic extent to a static extent, but must be equal; can't be determined until run time.

[explanation]

AM: Thank you, OK.

AM: How important is that [SFINAE]?

GR: We generally try to do SFINAE with constructors.

AM: OK.

AM: We could do compile-time matching on static extents

DS: Yes, but if ...

[JB arrives]

DS: But we're not requiring that; that could be QOI.

AM: Would it be allowed to compile?

GR: A conforming implementation could either fail at run time or fail to compile ... no wait ... in practice, it has no way of knowing whether the constructor call could be reached.

AM: I think we want Mandates ... if static extents don't mandate, that's Mandates.  If either dynamic, run time.

CE: Problem w/ mandate at compile time; if 3x3 or 6x6, if I'm a 3x3, construct static 3x3, if 6x6, then if statement.  If forbid compilation if it doesn't match ... it's like pulling a tile out.

AM: Should use if constexpr for this.

DS: This is based on 17, so we can use if constexpr.

CE: OK.

GR: If we're going the other way and say it has to compile if they don't match, would be difficult.

DS: If static extents don't match, should not compile.

AM: Might find it handy for users to have a compatible_extents function.

GR: That would need to go through LEWG.

AM: Could start by adding that as exposition / specification only, then ask [LEWG] if could be user facing.  In that case, would want to constrain on that, rather than mandates, to permit constructor to SFINAE.

GR: That seems borderline to me, but yes.  I can't imagine SFINAE finding another constructor that's OK if that condition fails, but yes.

GR: Here again, redundant Effects and Ensures [extents ctor from "other"]; drop Effects.

AM: Throws Nothing, and has a precondition, so it can't be noexcept. [OK]

-------

[ctor from dynamic_extents list]

GR: Has a precondition, so should be "Throws: Nothing" rather than noexcept, unless there is a strong engineering reason why it should be noexcept.

AM: dynamic_extents input parameter should have a different name, to avoid confusion with dynamic_extent.

[OK]

DS: dynamic_extent_args?

JB: non_static_extents?

VR: Question: If fold expression, why no parentheses around the whole thing?

DS: We need parentheses around the entire fold expression.

AM: Around every fold expression throughout the paper.

CE: Effects and Ensures are redundant again.

GR: Yes.

AM: "Initializes as if by {dynamic_extents...}" -- new wording, but saying "aggregate-initializes" feels too specific.

GR: OK.  Or we could dodge the problem, drop Effects, keep Ensures.

AM: Yes.

GR: Either way is fine.

AM: I would rather have Ensures than Effects.  Effects is [just] implementation advice.

GR: "For all i in zero to ..."  If we keep Ensures, it needs to say, "for all i in [0, SOMETHING) ..."

AM: When no dynamic extents, how does this partially order wrt default constructor?

DS: Because this is a templated one, does that make a difference?

[discussion]

GR: Let's just add a Constraint that rank_dynamic is not zero.  Then we don't have to worry about which one gets picked.

CE,DS: OK.

DS: Instead of just having it be variadic index type, could just name the first argument in the pack, so there is always at least one.

CE: Then you would still need rank_dynamic() > 0.

GR: I don't like that latter solution as much, since first element in pack isn't special in any way, would be misleading and make wording annoying.

CE: Just add it to the constraint.

AM: I don't want to have to force this thing to not exist, by writing extra code to silence it, if I don't have to.  I'd rather have a Note than a Constraint in that case, so that people won't write unnecessary code.  People might argue observability.

GR: Action item for you: Figure out if constraint is needed.

VR: The default ctor wins.

AM: OK, then just a Note is fine.

GR: If rank_dynamic is one, this would be an implicit conversion; this needs to be explicit.  Mark this as an unconditionally explicit constructor.  No value in having it be implicit for multiple arguments, and would be quite misleading, for multiple arguments.

GR: For std::array ctor: This implicitly requires the array to have size at least rank_dynamic, by virtue of subscripting into it, but nothing would forbid it being larger.  Is that intentional?

CE: No.  Should just be a different incarnation of the one above it.

AM: I had template parameters that are lowercase -- in this case, rank_dynamic.  This looks like an enum value.

[Prefer CamelCase.]

GR: Solve both problems at once: rank_dynamic should not be a parameter at all; just use the template parameter deduced from the std::array.

GR: Do you intent this to be implicit?

[No.  It should not be implicit.]

[Pass this by LEWG.]

[discussion of {thing, thing, thing} init.  If I say extents{thing, thing, thing}, which constructor?]

AM: {} does narrowing conversions ...

GR: I think one set of {} gives you IndexType..., and {{}} gives you the array<...> ctor.

DS: With concepts, could I constrain IndexType?

AM: It would simplify some of the constraints.  Narrowing conversion ... square brackets ...

DS: If we're going to go through that ...

AM: I would ask that in a larger group than this one.

GR: If things in curlies don't all have the same type ... If types don't match, it will fail to deduce index type.

DS: Could add deduction guide.

AM: Have you done a deduction guide for this class?  Implicit guides could be enough.

GR: If any of the implicit guides work at all, then they are probably broken.  The template parameters are intended to convey information that is not present in any of the constructors, at least any of the ones we've seen thsu far.

GR: Is there a way to disable implicit guides without adding explicit?

AM: No.  I think we talked about "= delete" in 17 but didn't do it.

GR: Homework for you: Figure out interaction with deduction guides.

JB: No, the two-curly-brace one can't deduce the std::array, because it also needs to deduce the index type.

CE: Is Effects better than Ensures in the array ctor?

GR: It has the same problem here...  "Initializes dynamic_extents_ with dynamic_extents" is fine.

AM: In terms of contracts, I would rather say Ensures rather than force how you get there, and let vendors decide how to get there.

GR: For array<IndexType, rank_dynamic> ctor, don't define D, since you're using it only once.

[Also, rename rank_dynamic template parameter.]

GR: For operator= this Requires is phrased differently than the corresponding constructor (that tends extents<OtherExtents...>).  It's not obvious to me if that is for an interesting reason.

CE: The other one is English; this one is code.

GR: And this is a logical or, whereas the wording for the other is if-then.

GR: Short-circuiting -- static_extent(r) is a run-time error to call with non-static r?

DR,CE: Yes.

GR: I think I prefer the English in this case (operator=).

CE: Effects and Ensures are redundant again (in operator=).

GR: Yes.

AM: Heterogeneous comparisons on extents?

DS,CE: Yes.

GR: Wording should be mostly identical to ctor.

CE: Yes.

---++ mdspan.extents.obs

CE: You can't be static constexpr and const.  (rank(), rank_dynamic()).

GR: This is a bool, but you're applying plus to it.  It's bad style.

JB: It's only recently that ++ on a bool was disallowed.

GR: I would say, make the Note the normative wording.  "The number of dynamic extents."  Another option is to return the size of the exposition-only array, but I prefer words here.

AM: Actually, that is the answer.  It returns the size of the exposition-only array.

GR: I'm OK with that, but I prefer having English.

CE: It's a loop if you do that.  We define length of the array based on the output of this function.

[This means, use English.]

JB: It (rank_dynamic() in std::array bound) doesn't compile if extents is not a template class.

GR: You're instantiating the class to get rank_dynamic(), that you use to define the array bound in the class.

JB: Both GCC and Clang accept it.

[Could define a stand-alone value.]

GR: If you don't do that, I would want a "see below" in the std::array bound, with a definition.

DS: Could we define rank_dynamic() inline?

GR: That wouldn't make a difference to this issue.  What I'm concerned about, is whether it's necessary to instantiate the class in order to call the rank_dynamic() better.

VR: I think it's OK.

GR: This is a case of we know what you mean.  I do like having the English instead of the fold expression, in rank_dynamic().

JB: It's definitely necessary to have the in-line definition of that routine, before you use it.

GR: If it's out of line, it definitely won't work, but the standard doesn't specify whether it's in line or out of line.

GR: Why does static_extent(r) return 1 if r is out of bounds?

CE: Users want to write generic code and don't want to do partial specialization.

GR: OK.  A 3-D mdspan is equiv to a 5-D with last two dims 1.

GR: Here again, "const".

GR: Has there been discussion on specifying dynamic_extents_ as an arry of the dynamic extents only, which is obviously the high-quality implementation, but for exposition purposes, it would be just as accurate to store all the extents -- and then you don't have to refer to DynamicRank in the wording?

CE: But then construction gets difficult.

GR: Does switching to just a postcondition help with that?

AM: If you're passing in the number of arguments, you're only passing in dynamic extents.

GR: OK, fair enough.

VR: static/dynamicextents being used in ... does not work in MSVC.  Don't know if it's a bug in MSVC.

CE: Perhaps we need to put in fold expression... should be "see below"

JB: What if you declare a static private constexpr variable?

GR: it's not sizeof(extents...) it's the number of elements ...

DS: [shows an alternative]

static constexpr size_t rank_dynamic_impl = (size_t(Indices == dynamic_rank) + ...); // exposition only

GR: No, that exposition-only member doesn't pay for itself, because it's only used in this one place.

CE: Is it really that bad to put the fold expression in the exposition-only ?

GR: I supposed...

AM: If you gave rank_dynamic_impl a _nice_ name, then it's a reasonable ...

CE: Capital RD since exposition only?

DS: We're using that value throughout.

GR: The nice name could be rank_dynamic.

CE: It needs to be a function.

GR: Could it be a value instead?

CE: Originally, long ago, it was an enum value.  Had thought the pattern was to make these things static constexpr functions.

AM: Could we inline it now?  Else would allocate storage for this thing.

GR: Inline is implied for static constexpr class members.  You might still need a definition if it's ODR-used.

JB: Don't use constexpr; use an enum.

AM: I don't want to go there.  Enums not useable in these context.

CE: We were discouraged from using enums.

GR: I'm pretty sure we don't need the enum hack any more, and I don't want to put new instances of it in standard documents.

JB: You should be fine (constexpr).

GR: Even putting the fold expression in the template parameter would be preferable to this (rank_dynamic_impl constexpr value).  Then, only one exposition-only member.

CE: Could define rank_dynamic() to return rank of array ...

GR: I still prefer the English.  Personally, I would say, don't do the function call in the std::array bound.  Either, do the fold expression, or see below.  Authors' choice.

DS,CE: We'll put the fold expression into the array.

GR: Please do it like you have it here, where you case bool to size_t.

MH: extent(r): "is true" in condition.

JB: MSVC does not accept the fold expression inside the declaration of the std::array.

GR: Do you know why?

JB: "... Extents parameter pack must be expanded in this context."  It works in Clang.

GR: Sounds more like an MSVC bug.

---++ mdspan.extents.compare

AM: Regarding operator== , I'm presuming == is preferable to <=> at this point?  Oh, it would be heterogeneous at this point, so could not be operator<=>.

GR: We're also not defining an order.

AM: Ah.

GR: For operator== would prefer English "and" to join the two bool expressions, to avoid confusion with for-all qualification.

AM: "lhs.extent(r) == rhs.extent(r)" (not "extents").

GR: Did you consider operator[], instead of calling it extent(r)?

CE: We didn't even have that discussion.

DS: We didn't want to overload operator[] ...

AM: Where this will be applied, there will be a lot of overloading.

GR: Fair.

---++ mdspan.layout

GR: 1: "which is a nested class template"

AM: Have we already defined layout mapping somewhere else?

CE: This is the first.

AM: I would start with bullet 3 before bullet 1.

CE: Bullet 2 is part of Bullet 3.

GR: Wait, a layout mapping and a layout mapping policy?

CE: Yes, it's a table that has both of them.

AM: I'd rather see things defined in order of use.

CE: Are 3 and 4 the only clauses we need?

AM: I would like to see the order 3, 1, 2.

GR: So far, only the first line in the table deals with the layout mapping policy.

MH: Perhaps then we should separate that out?

GR: Yes.  Would perhaps be silly to have a table for one line, but no other way to express.

AM: You move it up into Bullet 4.

GR: I would move these (1 & 2) down, begin w/ paras 3 and 4, and after table, say "a layout mapping policy" ...

AM: MP is only used once in this line [of the table] we can get rid of.  "A class MP is a layout mapping policy if MP::template mapping<E> meets the requirements of a layout mapping .. where E is any specialization of extents."

GR: Can an instance of extents ever fail to define a multidimensional index space?

MH: You can't give it negative dynamic extents that aren't dynamic_extent; that's ill formed.

GR: If there are ways that it could define a multidimensional index space, then we should fix extents so it can't.

CE: That's just color commentary.

GR: Then let's drop it.

[OK]

GR: What does "denotes a multi-index in e" mean?

[Pablo Halbern arrives]

GR: What this is trying to say, is that i and j in the wording below, are packs of integer types, where each element of the pack is within the corresponding extent.

GR: Does this library define an operation that tests whether a multi-index is in an extent?

CE: We have clauses.

GR: Is there a query a programmer can write?

CE: There's no function defined to do that.  All the definitions are there, but we didn't define a function for it.

GR: That would be useful for simplifying wording.

DS: We do have an internal function like that for bounds checking.  We could expose that, or define it as exposition only.

GR: Is there anywhere else in the wording where we talk about that?

CE: Down in the Expects ...

[skipping down to look at that wording]

GR: Yes, e.g., operator()(Indices... i).

GR: I think we should factor out some reusable way of wording that, whether it's a library operation or a wording macro.

CE: We could put the wording macro here.

GR: We probably would like to put that in extents.

GR: Is there any particular reason why you didn't expose that to users?

MH: Typical user idiom is to query the mdspan's dimensions, not to pass in an entire multi-index.

AM: I'm a huge fan of functions that help specify a feature; people tend to find them useful anyway.  Analogy with test-driven development.

GR: Anybody who writes wrapper logic that takes an mdspan and indices separately might want to check that the indices are valid before they use the indices.  I would encourage you to add that, probably a member of extents.  Would need to go to LEWG for that.

[OK]

GR: Last bullet before table: Is this also color? or does it impose an add'l constrain t on m?

CE: Color.

GR: I say strike it; too hard to tell if color or constraint.

GR: "m.extents()" -- "Returns: e" -- not a constraint [VR leaves] on the function call, but on e.  We could equivalently said, up above in bullet points, that object is returned by "m.extents()", except that we would have to shuffle order to define m first.  Do we use "e" anywhere in the table?  Yes.

CE: If we had the aforementioned function, we could replace "in E" with that function.

GR: Do we do this kind of thing in other tables, where individual entries of the table determine constraints on the placeholders that then affect other entries in the table?

AM: I think we do it in containers.

GR: OK.

GR: "Requires: 0 <= m(i...)." is a precondition, but this is an Ensures on operator() of m ... not really an Ensures ...

CE: operator() on m is required to return nonnegative

GR: ... it's still ... this is a Requires element; those go on functions, not on objects or types.  I want to read this as a precondition on the caller of m.  The phrasing makes me think it's a precondition on the caller of m(i...), but it's a precondition on m.  Have we defined what a mapping of a multi-index is?

CE: We have the color commentary up there.  It returns a nonnegative value.  Could we put that in Returns?  "Returns a nonnegative value"?

DS: Mappings are not guaranteed to be unique.

GR: "Returns a nonnegative value."

[OK]

GR: m.required_span_size(): E defines an empty space if one of its extents is zero?

CE: Correct.

GR: It might be better to be more clear about that.  I'm wondering if extents should have a .empty() method, but "if e.extent(i) is zero for any i in whatever" could suffice.

CE: After the extents section, but before the mapping section ... "defines a multidimensional index space, of the elements satisfying this condition" ...

GR: Probably part of extents ... maybe right after the extents synopsis.  That makes sense.

JB: For is_unique(), I'm confused why the ... are there.

GR: They are parameter packs, i... and j... -- they don't need parens, but is "m(i)..."  OK?

DS: should be "m(i...)" not "m(i)...".

GR: You need to quantify that over all i, right?

CE: Yes.  For all members of the multidimensional index space defined by E.

GR: Rather than talking about this set, ... you have two quantifiers here: For all "flattened indices" / "values in the codomain", there exists a value in the domain, much that m(i...) equals that value.  I don't see a clean way to simplify that.

DS: Can simplify with math syntax.

GR: This would be a good place to use math syntax.

DS: Unambiguous with math syntax.

GR: Does this sigma need a subscript?

AM: What's benefit of using sigma instead of just a letter?

MH: Avoid confusion with S.

DS: It breaks my web browser.

CE: Let's take sigma out.

DS: Math font in general wasn't rendering on all browsers.

GR: Stick to the Roman alphabet if possible, and avoid subscripts unless needed.

GR: Don't even talk about "there exists" (is_strided()) -- where does r come from?

CE: It's the ordinate -- see above.

GR: Subscript implicitly quantifies: a set of integers for r in the possible ranks?

CE,DS: Yes.

GR: So you don't need sigma, but you still need r.  Should be argument to m.stride rather than subscript.

DS: Yes.

CE: Definition of m.stride(r) depends on m.is_strided() definition.

GR: Is this something you expect user code to call?

DS,CE: Yes.

GR: We could define a ... we're designing here, but this member is_strided is redundant with the existence of stride, right?

CE: No.

MH: If is_strided is false, does stride(r) ever make sense for any r?

DS: No.

CE: ... is_always_strided etc. are for metaprogramming.

[SFINAE to get rid of stride(r)?]

[Need to rephrase.  This Table is still not clear.]

DS: Now that we have consteval, would those become consteval?

MH: constexpr still correct.

GR: We have terrible wording tech around constexpr in the library; leave well enough alone now.

[PH leaves.  CH left a little while ago.]

JB: If routine doesn't take any parameters, constexpr is appropriate.

[...]

CE: m.required_span_size(): Tight upper bound.  Used for allocating a multidimensional array.

---++ layout_left

GR: Note (para 1) seems superfluous.

CE: Hinted by "Thus,"

GR: Gives a strided layout mapping?

DS,CE: This is not a strided layout mapping; we have one of those.

GR: So layout_left gives/has a strided layout mapping?

CE: Yes, but there is a layout_stride, which is something different.

GR: This is _a_ strided layout, yes?

CE: Yes.  Gives a layout mapping that is always strided?

CE: Gives a strided layout mapping?

DS,CE: Concern about overloading term

GR: But it's not overloading

CE: Fear user confusion.

GR: You're worried that people might be confused between "a strided layout" and "layout_stride".

CE: yes.

GR: Then that's a problem with the name of layout_stride.  layout_left _is_ a strided layout.  If you're afraid to describe it that way ...

GR: Is any strided layout representable using layout_stride?

CE: Yes, it is the superset of all strided layouts.

GR: In particular, you could construct an instance of layout_stride that could represent an instance of layout_left?

CE: Yes, just less efficiently.

GR: Analogous to std::swap being a function and we don't worry about confusion with std::function.

DS: OK, "is a strided layout."

DS: We got only to the first couple things in layout_left.  Will get a paper for post-meeting that includes these updates and a pass at other wording.  Marshall wants to do papers on the reflector.  This is a candidate.

[Someone asked if mdspan has a reference implementation; it does.]

---++ Comments

%COMMENT%
```

**Comment by mhoemmen** [comment]
Retitled after consultation with @dsunder .

**Comment by dhollman** [comment]
Notes:

- AFAIK a move constructor is *never* QOI. There needs to be an implementation-agnostic response to `std::is_move_constructible_v`.  We very definitely should not strike the move constructor from extents. I can't think of any analog in the standard for this sort of thing; can anyone else? If this happened, you could have one implementation where `std::swap(ex1, ex2)` works fine, and then in another implementation it doesn't even compile. That's not how standards work, AFAICT.
- `this->extent(r)` would be inconsistent with how all of the rest of the standard does things.  See, e.g., `span`.  We just want `extent(r)` there. (Oh, just got to the point where GR withdrew this comment, never mind)
- Compile-time matching of static extents can't be QOI for the same reason as the move constructor. `is_convertible_v` should be the same across implementations.
- The comment about value-initializes instead of aggregate-initializes: since `dynamic_extents_` is a `std::array`, value-initialization doesn't do what you want for built-in types.  This should be aggregate-initializes.

More comments to come...

**Comment by dhollman** [comment]
More notes:

- GR's comment that the effects clause is redundant in the constructors is incorrect AFAICT.  Those clauses are needed to make the exposition-only variable meaningful in the rest of the wording.  They should remain in there. (I mimicked several other uses of exposition-only variables in the standard in this respect. I think there's a comment in the bikeshed source listing what I mimicked).
- same comments as above with `dynamic_extents...` ctor.  I'm pretty sure I did my homework on these, so if I'm wrong I'd like to see more than a passing comment as to why I'm wrong.
- I think the `rank_dynamic() != 0` constraint on the `dynamic_extents...` constructor is unnecessary, since the `is_convertible_v` fold expression will be `false` in that case.
- I think there's a lot of confusing arising in the `std::array` constructor from the fact that somehow we ended up with a deduced `rank_dynamic` template parameter.  `rank_dynamic` should have been non-deduced and fixed to the value`rank_dynamic()`, just as in the exposition-only parameter. I don't know if this was my typo or got added later, but hopefully this clears things up.

**Comment by dhollman** [comment]
More:

- Deduction guides are not an issue; the template parameters don't participate in any of the constructors, so none of them should produce implicit deduction guides and there's no way to produce explicit deduction guides from any of the constructors since the information about the sizes of the static extents simply isn't there.
- On the use of `rank_dynamic()` in the type of the exposition only member—if it helps, I tried all of the alternatives discussed in the notes, and ended up with this one (I guess I didn't test MSVC...); seems like the room settled on the same thing?

On layout:

- The "defining a multidimensional index space." piece is there to indicate that references to multidimensional indices in the table that follows are part of that space.  I went back and forth on how to clarify that, and landed on that phrasing.
- The last bullet before the table is important for tying `i...` to `e`.  Every other way I tried to do this was less clear.  We can't just strike this unless that is made clear elsewhere.

**Comment by mhoemmen** [comment]
Pending #90 changes w/ @hcedwar .
